packages/sitecore-jss-angular/src/components/placeholder.component.ts

```diff
@@ -166,15 +175,90 @@ export class PlaceholderComponent implements OnInit, OnChanges, DoCheck, OnDestr
         }
       }
     }
+    this.placeholderData = this.renderings || this.getPlaceholder() || [];
+    this.contextSubscription = this.jssState.state.subscribe(({ sitecore }) => {
+      this.editMode = sitecore?.context.editMode;
+      this.metadataMode = this.editMode === EditMode.Metadata;
+    });
   }
 
+  getPlaceholder() {
+    let phName = this.name?.slice() || '';
+    /**
+     * Process (SXA) dynamic placeholders
+     * Find and replace the matching dynamic placeholder e.g 'nameOfContainer-{*}' with the requested e.g. 'nameOfContainer-1'.
+     * For Metadata EditMode, we need to keep the raw placeholder name in place.
+     */
+    this.rendering?.placeholders &&
+      Object.keys(this.rendering.placeholders).forEach((placeholder) => {
+        const patternPlaceholder = isDynamicPlaceholder(placeholder)
+          ? getDynamicPlaceholderPattern(placeholder)
+          : null;
+        if (patternPlaceholder && patternPlaceholder.test(phName)) {
+          if (this.editMode === EditMode.Metadata) {
+            phName = placeholder;
+          } else {
+            this.rendering.placeholders![phName] = this.rendering.placeholders![placeholder];
+            delete this.rendering.placeholders![placeholder];
+          }
+        }
+      });
+
+    if (
+      this.rendering &&
+      this.rendering.placeholders &&
+      Object.keys(this.rendering.placeholders).length > 0
+    ) {
+      return this.rendering.placeholders[phName];
+    }
+    return null;
+  }
+
+  getCodeBlockId = (kind: string, renderingId?: string): string | undefined => {
```

Comment:
I'm wondering if it's better to create a `PlaceholderMetadata` class that will be extended by `PlaceholderComponent` class. In this case we can hide some initialization logic, variables, methods. Currently, everything is mixed here

Reply:
I've had trouble for some time previously navigating between Placeholder and PlaceholderCommon in React when debugging. 
And for all the mixing in, I believe this to be a better, more visible approach, with a single placeholder component. What you see in it is what you get, without splitting implementation between multiple files.

Reply:
And I'm still wary, not sure if this is a good idea. Angular has a lot of quirks around its template. The whole functionality resolves around the private view reference. I did a quick check - compiler doesn't like when view elements are defined on both levels. I think we would need the child to have its own view element in template ob child, but handle it the same as parent most of the time in _render() method. 
Or alternatively move placeholder code blocks renders from template into code which I also would like to avoid - this the only visual representation of code blocks in template, the only indication this is how they will be used.

Reply:
Ok, let's leave it
My main request was to analyze if it's possible to improve it